src/lsp.ts

```diff
@@ -0,0 +1,180 @@
+import * as Admzip from "adm-zip";
+import * as which from "which";
+import * as https from "https";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { Executable, LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  config: WorkspaceConfiguration;
+  ls: LanguageClient | null = null;
+  executableName: string;
+  context: ExtensionContext;
+  private clientOptions: LanguageClientOptions = {
+    documentSelector: ["meson", { "scheme": "file", language: "meson" }]
+  };
+  abstract repoURL: string;
+  abstract get debugExe(): Executable
+  abstract get runExe(): Executable
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName;
+    this.context = context;
+    this.config = workspace.getConfiguration("mesonbuild");
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer();
+  }
+
+  abstract supportsSystem(os: string, arch: string): boolean
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
+    this.dispose();
+
+    if (this.languageServerPath == null)
+      return;
+    this.startLanguageServer();
+  }
+
+  private startLanguageServer() {
+    const serverOptions: ServerOptions = {
+      run: this.runExe,
+      debug: this.debugExe,
+      transport: TransportKind.stdio
+    };
+
+    this.ls = new LanguageClient("Swift-MesonLSP", "Meson Language Server", serverOptions, this.clientOptions, true);
+    this.ls.start();
+  }
+
+  canDownloadLanguageServer(): boolean {
+    return this.downloadInfo != null && this.findLanguageServer() == null
+  }
+
+  get languageServerPath(): string | null {
+    return this.config.languageServerPath || this.findLanguageServer() || which.sync(this.executableName, { nothrow: true })
```

Comment:
Is it possible to just resolve the language server path in the constructor?

Reply:
No, as e.g. the PATH can change (Drop binary into /usr/bin) and findLanguageServer() returns the path to the downloaded executable => Not cachable if I understand correctly

Reply:
Ok, I think we should just resolve the path in the constructor, and add a command to "restart the language server" similar to clangd. Then a user can change the path of the language server all they want.